### PR TITLE
[shortfin] Add end-to-end cli for testing / executing prefill

### PR DESCRIPTION
**Pull request by rsuderman**
Updated for what the outgoing prefill looks like. This included validation with llama 2 that a reasonable sequent (one two three four) correctly predicted preceding tokens.

This should support generic export / execution of gguf prefill however there still exists issues with the newly added decoder path.

### CODE REVIEW

File: shortfin/shortfin/llm/impl/service_v1_cli.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+
+import asyncio
```
Comment (ScottTodd) [inline shortfin/shortfin/llm/impl/service_v1_cli.py:2]: nit: add copyright header please

File: shortfin/shortfin/llm/impl/service_v1_cli.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+from shortfin.llm.service import GenerateRequest
+
+def setup(vmfb_path, config_path, gguf_path):
+    from iree.runtime._binding import disable_leak_checker  # type: ignore
+    model_params = ModelParams.load_json(config_path)
+
+    cache_params = CacheParams(
+        model=model_params, device_block_count=128, block_pos_stride=16)
+
+    disable_leak_checker()
```
Comment (ScottTodd) [inline shortfin/shortfin/llm/impl/service_v1_cli.py:36]: 👀  leaks?

File: shortfin/shortfin/llm/impl/service_v1_cli.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    return mapped.asarray(
+        value.shape,
+        HalElementType.map_to_dtype(value.element_type))
+
+async def main(argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--tokenizer")
+    parser.add_argument("--config")
+    parser.add_argument("--vmfb")
+    parser.add_argument("--gguf")
```
Comment (ScottTodd) [inline shortfin/shortfin/llm/impl/service_v1_cli.py:61]: Document these? (They look pretty self-explanatory to me right now, but docstrings are easy enough to add and will help others in the future)

File: shortfin/shortfin/llm/impl/service_v1_cli.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    hf_path = parsed.tokenizer
+    config_path = parsed.config
+    vmfb_path = parsed.vmfb
+    gguf_path = parsed.gguf
+
+    service = setup(vmfb_path, config_path, gguf_path)
+    tokenizer = LlamaTokenizer.from_pretrained(hf_path)
+    state = service.start()
+
+    for line in ["one two three four"]:
```
Comment (ScottTodd) [inline shortfin/shortfin/llm/impl/service_v1_cli.py:73]: Want to expose this as an argument, or will the service not need that?

File: shortfin/shortfin/llm/impl/service_v1.py [hunk 5]
```diff
@@ -315,3 +361,139 @@ async def prefill(self) -> TimelineGuarded[HalBufferView]:
...
+        (
+            decode_attn_block_indices_host,
+            decode_attn_block_indices_device,
+        ) = resources.acquire_transfer_buffer(service.block_indices_pool).h2d_array(
+            cb, [bs, max_attn_blocks_length], HalElementType.SINT_64, fill_value=0
+        )
+
+        # Populate host buffers for each sequence.
+        for i in range(len(sequences)):
+            break
```
Comment (ScottTodd) [inline shortfin/shortfin/llm/impl/service_v1.py:456]: 🤔 immediately break? add TODO comment with some context on what you were debugging maybe?

**Review by ScottTodd — APPROVED** [review]
Looks good to land and continue iterating on. A few small comments.

File: shortfin/shortfin/llm/impl/service_v1.py (hunk 5 above)
Comment (rsuderman) [reply shortfin/shortfin/llm/impl/service_v1.py]: Oops, debug code for tracking down issues in the decoder implementation. Fixed.

File: shortfin/shortfin/llm/impl/service_v1_cli.py (hunk 4 above)
Comment (rsuderman) [reply shortfin/shortfin/llm/impl/service_v1_cli.py:81]: Eventually yes, for now I wanted to leave it as a fixed string for testing.

File: shortfin/shortfin/llm/impl/service_v1_cli.py (hunk 2 above)
Comment (rsuderman) [reply shortfin/shortfin/llm/impl/service_v1_cli.py:44]: Unfortunately yes - I think a decent chunk of this is how python releases resources. There may be a better way to track this down but its replicated on the existing service work.

File: shortfin/shortfin/llm/impl/service_v1_cli.py (hunk 3 above)
Comment (rsuderman) [reply shortfin/shortfin/llm/impl/service_v1_cli.py]: done.

File: shortfin/shortfin/llm/impl/service_v1_cli.py (hunk 1 above)
Comment (rsuderman) [reply shortfin/shortfin/llm/impl/service_v1_cli.py:7]: Done.

**Pull request merged**